test/tools/test_tools_memories.py

```diff
@@ -58,3 +62,412 @@ def test_sopel_identifier_memory_channel_str():
     assert memory[channel] == test_value
     assert memory['#adminchannel'] == test_value
     assert memory['#AdminChannel'] == test_value
+
+
+def test_sopel_identifier_memory_setdefault():
+    memory = memories.SopelIdentifierMemory()
+    assert 'foo' not in memory
+
+    memory.setdefault('Foo')
+    assert len(memory) == 1
+    assert 'foo' in memory
+    assert 'fOO' in memory
+    assert memory['fOo'] is None
+
+    memory.setdefault('bAr', 'DEFAULT')
+    assert len(memory) == 2
+    assert 'FoO' in memory
+    assert 'BaR' in memory
+    assert memory['Bar'] == 'DEFAULT'
+
+    assert memory.setdefault('baR', 'different') == 'DEFAULT'
+    assert len(memory) == 2  # still, that is
+
+
+def test_sopel_identifier_memory_copy():
+    memory = memories.SopelIdentifierMemory()
+    memory['SomeCamelCase'] = True
+    memory['loweronly'] = False
+    assert len(memory) == 2
+    assert type(memory).__name__ == 'SopelIdentifierMemory'
+
+    copied = memory.copy()
+    assert len(copied) == 2
+    assert 'SomeCamelCase' in copied
+    assert 'loweronly' in copied
+    assert type(copied).__name__ == 'SopelIdentifierMemory'
+
+
+def test_sopel_identifier_memory_get():
+    memory = memories.SopelIdentifierMemory()
+    memory['SomeCamelCase'] = True
+    memory['loweronly'] = False
+    assert len(memory) == 2
+
+    # verify key-exists behavior w/implicit default
+    got = memory.get('somecamelcase')
+    assert got is True
+    del got
+    got = memory.get('LowerOnly')
+    assert got is False
+    del got
+
+    # verify key-exists behavior w/explicit default of None
+    got = memory.get('somecamelcase', None)
+    assert got is True
+    del got
+    got = memory.get('LowerOnly', None)
+    assert got is False
+    del got
+
+    # verify key-exists behavior w/explicit "real" default value
+    got = memory.get('somecamelcase', 'DEFAULT')
+    assert got is True
+    del got
+    got = memory.get('LowerOnly', 'DEFAULT')
+    assert got is False
+    del got
+
+    # verify key-missing behavior w/implicit default
+    assert 'missing_key' not in memory
+    got = memory.get('missing_key')
+    assert got is None
+    del got
+
+    # verify key-missing behavior w/explicit default of None
+    assert 'missing_key' not in memory
+    got = memory.get('missing_key', None)
+    assert got is None
+    del got
+
+    # verify key-missing behavior w/explicit "real" default value
+    assert 'missing_key' not in memory
+    got = memory.get('missing_key', 'DEFAULT')
+    assert got == 'DEFAULT'
```

Comment:
I hate the `del got` sprinkled over this test, so I had a go at rewriting. It could be simplified further if not for the mixed case in the keys (i.e. as assign `'SomeCamelCase'` but we test `somecamelcase`), maybe it would be better to write a test dedicated to testing `get()`'s behavior over case, or maybe this test itself should be several (parametrized?) tests.

Anyway, just scratching my own itches here.

Note that I dropped the repetition of `assert 'missing_key' not in memory` in the last clause, which I think is only necessary if we're trying to test that `get()` does not mutate the memory object (which is maybe another point in favor of breaking this into multiple tests)

```suggestion
def test_sopel_identifier_memory_get():
    memory = memories.SopelIdentifierMemory()
    memory['SomeCamelCase'] = True
    memory['loweronly'] = False
    assert len(memory) == 2

    # This could be used for the above, too, except for the
    # case change. That might be something that could be
    # tested elsewhere, though…
    pairs = [
        ("somecamelcase", True),
        ("LowerOnly", False),
    ]

    for key, expected in pairs:
        # verify key-exists behavior w/implicit default
        assert memory.get(key) == expected

        # verify key-exists behavior w/explicit default of None
        assert memory.get(key, None) == expected

        # verify key-exists behavior w/explicit "real" default value
        assert memory.get(key, 'DEFAULT') == expected

    # verify key-missing behavior w/implicit default
    assert 'missing_key' not in memory
    assert memory.get('missing_key') is None
    # verify key-missing behavior w/explicit default of None
    assert memory.get('missing_key', None) is None
    # verify key-missing behavior w/explicit "real" default value
    assert memory.get('missing_key', 'DEFAULT') == 'DEFAULT'
```

Reply:
I'm going to adapt some, but perhaps not all, of the ideas from this rewrite for both the `get()` and `pop()` tests, and give you co-author credit for the inspiration. Stay tuned!

> maybe it would be better to write a test dedicated to testing `get()`'s behavior over case

This basically _is_ said test dedicated to testing `get()`'s behavior over case. The problem this patch fixes is that accessing a key with arbitrary casing through `__getitem__()` worked properly, but `get()` didn't.

We don't _really_ need to check that `get()` actually gets something, per se, assuming we trust the Python runtime's test suite to validate the behavior of `super().get()`. What we need to check is whether our overridden `get()` modifies the input correctly, to retrieve values in those cases where `dict.get()` itself wouldn't. That's why I so often create or assign values using one casing of a key, but retrieve them with a different variation: It checks the behavior _we_ control, which is key transformation.

(But the tests do check the whole range of `get()` behaviors as documented in Python's docs, for completeness. Just in case a future code change breaks something else. Regressions suck.)